### PR TITLE
Refactor comment algorithm and improve newline/spaces detection

### PR DESCRIPTION
**Pull request by jlongster**
I'm posting this if anyone wants to look at it early. This fixes a bunch of major bugs in comments, but there are still plenty to be ironed out. 

An example that fails:

```js
var x = {
  foo: 5,
  bar: 6 // sdfsfd
}
```

It adds a newline:

```js
var x = {
  foo: 5,
  bar: 6 // sdfsfd

};
```

But at least the number of times we emit invalid JS code should be less. I will add a lot more comments too.

See the tests, there are still more things to fix. Right now if there are inline comments, it tends to force things to be broken up because otherwise what would we do? We could try to convert them into block comments if we collapse, but that seems a lot harder.

### CODE REVIEW

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -133,8 +128,8 @@ export type BuckWebSocketMessage = {
   } | { type: \"BuildProgressUpdated\", progressValue: number } | { type: \"BuildFinished\", exitCode: number } | { type: \"BuildStarted\" } | { type: \"ParseStarted\" } | { type: \"ParseFinished\" } | { type: \"RunStarted\" } | { type: \"RunComplete\" };
 
 // Missing one level of indentation because of the comment
-const rootEpic = (actions, store) => // Log errors and continue.
-combineEpics(...epics)(actions, store).catch((err, stream) => {
+const rootEpic = (actions, store) => combineEpics(...epics)(actions, store)
+// Log errors and continue..catch((err, stream) => {
```
Comment (vjeux) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: this is invalid :)

File: tests/destructuring/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -197,26 +191,17 @@ obj_rest({ p: {} });
 obj_rest({ p: { q: 0, r: null } });
 
 function obj_prop_annot(
-  {
-    // error: boolean ~> string
-    p = true
-  }: { p: string } = {
-    // error: number ~> string
-    p: 0
+  { p = true // error: boolean ~> string }: { p: string } = {
```
Comment (vjeux) [inline tests/destructuring/__snapshots__/jsfmt.spec.js.snap]: invalid too

File: tests/comments/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (jlongster) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Yep, there's definitely some regressions

File: tests/destructuring/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (jlongster) [inline tests/destructuring/__snapshots__/jsfmt.spec.js.snap]: Yeah, I have a fix for this one coming later

**Comment by jlongster** [comment]
Unfortunately this doens't fix too many of the existing issues, but it does fix a few bad bugs. The main thing it does is starts providing a framework for dealing with comments, i.e. figuring out patterns for moving them around "after the fact". I'll build a few more primitives and I think we'll be getting somewhere.

The thing that's really hard about comments is that are added after a node is printed. That means they exist outside of any indentation, commas, etc places by printing the AST node. This makes it very tricky to do things line inline comments, and make sure the comment is printed exactly on the same line it was in the original source. Maybe it should exist on the same line as an expression, but the expression has already added a `hardline`, and we can't easily remove it because it exists inside other groups/concats/etc. Potentially we could traverse the document and strip the last newline but that's going to have a perf hit.

Additionally what happens if you have a comment like this?

```js
var x = {
  foo: 1 // hello
};
```

And the printer tried to collapse the object? We either need to turn it into a block comment (not ideal) or force it to stay expanded. Right now it forces it to stay expanded by keeping the newline at the end of the comment as a hard line, but comments can be anywhere and that means we're allowing the user to force a newline at any point in the code, which must produce bad code sometimes (there are places we just don't expect newlines).

Anyway, more research is needed, but I'm feeling like I understand the problem a bit clearer now.

**Comment by jlongster** [comment]
An easy example where inserting a hard line breaks things. I'm sure there are many more:

```js
var x = // foo
  5;
```

Outputs:

```js
var x = // foo
5;
```

This parses fine (I think) but I'm not sure we want to allow them to be anywhere.

**Comment by vjeux** [comment]
For end of line comment what about the following:

Add a new type in the IR: eolComment(comment)
In the measure phase, return comment.length for this node
In the print phase, add it to a buffer and whenever you print a \n then print all the buffered end of line comments.

**Comment by vjeux** [comment]
In case there are more than one comment to be merged, what we can do is to turn them into inline comments. When we put them in the buffer we know the current position so it should be easy to do.

In your examples above,

```js
var x = { foo: 1 }; // comment

var x = {
  foo: 1, // comment 1
  foo: 2, // comment 2
};

// would turn into

var x = { foo: 1 /* comment 1 */, bar: 2 /* comment 2 */ };

var x = 1; // comment
```

Because we still compute the length of the comments in the breaking phase, they are likely going to be kept expanded as the content + comments are likely going to be > 80 columns

**Comment by jlongster** [comment]
I'm hesitant about converting inline comments (when I say inline comments, I mean line comments at the end of line but in the middle of an expression) to block comments (`/*` and `*/`) because without reliably turning them back when the expression is expanded, it's annoying for the user to manually turn them back into line comments if they want. Also block comments are quite a different style. My gut feeling is it's better to force the expression to break if possible, otherwise we move the comment up if it's a line comment (which means disallowing putting line comments in some really weird places).

Adding some new primitives to handle comments sound promising. I've had that idea in the back of my mind but wanted to see how far we could get without it. Once we figure out the rules we want we can implement that and see how it goes. If we force an expression to be expanded and also have the ability to put a comment at the end of a line (no matter where it's places in the document), that might be really powerful.

At this point I'd love to talk to @jordewalke. These are the decisions I'm sure he's already gone through so we should follow his lead.

**Comment by vjeux** [comment]
For turning them back on: if we are going to print an inline comment at the end of the line, then make it a end of line block comment.

Also, I would only suggest turning eol comments into inline ones if there are more than 2 in a line, which is going to be an unlikely event.

**Comment by jlongster** [comment]
> Also, I would only suggest turning eol comments into inline ones if there are more than 2 in a line, which is going to be an unlikely event.

What do you mean by this? There cannot be more than 2 eol comments in a line, given that you can only put one at the end of the line: :)

```js
var x = 5; // end of line comment
```

**Comment by vjeux** [comment]
```js
var x = {
  a: 1 // comment
};

-->

var x = { a: 1 }; // comment
```

but

```js
var x = {
  a: 1, // comment
  b: 2, // comment
};

-->

var x = { a: 1 /* comment */, b: 2 /* comment */ };
```

**Comment by jlongster** [comment]
Once I get #440 I'll work on this some more.

**Comment by jlongster** [comment]
Lots of code has changed since this so I'm closing and will open a new PR.

**Pull request closed without merging**